src/sources/prometheus/remote_write.rs

```diff
@@ -0,0 +1,270 @@
+        PrometheusRemoteWriteSnapError,
+    },
+    prometheus::{proto, METRIC_NAME_LABEL},
+    shutdown::ShutdownSignal,
+    sources::{
+        self,
+        util::{ErrorMessage, HttpSource, HttpSourceAuthConfig},
+    },
+    tls::TlsConfig,
+    Event, Pipeline,
+};
+use bytes::Bytes;
+use chrono::{DateTime, LocalResult, TimeZone, Utc};
+use prost::Message;
+use serde::{Deserialize, Serialize};
+use std::{
+    collections::{BTreeMap, HashMap},
+    net::SocketAddr,
+};
+use warp::http::{HeaderMap, StatusCode};
+
+const SOURCE_NAME: &str = "prometheus_remote_write";
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct PrometheusRemoteWriteConfig {
+    address: SocketAddr,
+
+    tls: Option<TlsConfig>,
+
+    auth: Option<HttpSourceAuthConfig>,
+}
+
+inventory::submit! {
+    SourceDescription::new::<PrometheusRemoteWriteConfig>(SOURCE_NAME)
+}
+
+impl GenerateConfig for PrometheusRemoteWriteConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            address: "127.0.0.1:9090".parse().unwrap(),
+            tls: None,
+            auth: None,
+        })
+        .unwrap()
+    }
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "prometheus_remote_write")]
+impl SourceConfig for PrometheusRemoteWriteConfig {
+    async fn build(
+        &self,
+        _name: &str,
+        _globals: &GlobalOptions,
+        shutdown: ShutdownSignal,
+        out: Pipeline,
+    ) -> crate::Result<sources::Source> {
+        let source = RemoteWriteSource {
+            decompressor: snap::raw::Decoder::new(),
+        };
+        source.run(self.address, "", &self.tls, &self.auth, out, shutdown)
+    }
+
+    fn output_type(&self) -> crate::config::DataType {
+        config::DataType::Metric
+    }
+
+    fn source_type(&self) -> &'static str {
+        SOURCE_NAME
+    }
+}
+
+#[derive(Clone)]
+struct RemoteWriteSource {
+    decompressor: snap::raw::Decoder,
+}
+
+impl RemoteWriteSource {
+    fn decode_body(&self, body: Bytes) -> Result<Vec<Event>, ErrorMessage> {
+        let body = self
+            .decompressor
+            .clone()
+            .decompress_vec(&body)
+            .map_err(|error| {
+                emit!(PrometheusRemoteWriteSnapError {
+                    error: error.clone()
+                });
+                ErrorMessage::new(
+                    StatusCode::BAD_REQUEST,
+                    format!("Could not decompress write request: {}", error),
+                )
+            })?;
+        let request = proto::WriteRequest::decode(Bytes::from(body)).map_err(|error| {
+            emit!(PrometheusRemoteWriteParseError {
+                error: error.clone()
+            });
+            ErrorMessage::new(
+                StatusCode::BAD_REQUEST,
+                format!("Could not decode write request: {}", error),
+            )
+        })?;
+        Ok(decode_request(request))
+    }
+}
+
+impl HttpSource for RemoteWriteSource {
+    fn build_event(
```

Comment:
We could do this as a follow-up, but we've started adding config to other HTTP-based sources to optionally assign query parameters as fields on the events:

https://github.com/timberio/vector/blob/73a71363ea9115843f503801cac1f76bac752f19/src/sources/logplex.rs#L55-L65

I could see extending that to here. It can be useful to allow custom labeling by users by adding query parameters to the endpoint that points to the remote write source.

I could also see supporting setting of HTTP headers as fields similar to the vanilla HTTP source. It might be nice to support those features uniformly for HTTP-based sources.

Reply:
It sounds useful, but would be best as a follow-up so we can finish this up.